pkg/controllers/grafana-operator/controller.go

```diff
@@ -151,108 +158,161 @@ func RegisterWithManager(mgr controllerruntime.Manager) error {
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcileError(err)
 	}
+
+	// create
 	if errors.IsNotFound(err) {
-		r.logger.Info("Creating Grafana Operator OperatorGroup")
-		return r.k8sClient.Create(ctx, NewOperatorGroup())
+		r.logger.Info("Creating Grafana OperatorGroup")
+		err := r.k8sClient.Create(ctx, NewOperatorGroup())
+		return creationResult(5*time.Second, err)
 	}
 
-	r.logger.Error(err, "error reconciling operator group")
-	return err
+	// update
+	r.logger.Info("Updating Grafana OperatorGroup")
+	desired := NewOperatorGroup().Spec
+	operatorGroup.Spec = desired
+	err = r.k8sClient.Update(ctx, &operatorGroup)
+	return updationResult(err)
 }
 
-func (r *reconciler) reconcileSubscription(ctx context.Context) error {
+func (r *reconciler) reconcileSubscription(ctx context.Context) reconcileResult {
 	key := types.NamespacedName{
 		Name:      subscriptionName,
 		Namespace: Namespace,
 	}
 	var subscription v1alpha1.Subscription
 	err := r.k8sClient.Get(ctx, key, &subscription)
-	if err == nil {
-		r.logger.Info("Updating Grafana Operator Subscription")
-		subscription.Spec = NewSubscription().Spec
-		return r.k8sClient.Update(ctx, &subscription)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcileError(err)
 	}
+
+	// create
 	if errors.IsNotFound(err) {
 		r.logger.Info("Creating Grafana Operator Subscription")
-		return r.k8sClient.Create(ctx, NewSubscription())
+		err := r.k8sClient.Create(ctx, NewSubscription())
+		return creationResult(15*time.Second, err)
 	}
 
-	return err
+	r.logger.Info("Updating Grafana Operator Subscription")
+	subscription.Spec = NewSubscription().Spec
+	err = r.k8sClient.Update(ctx, &subscription)
+	return updationResult(err)
 }
 
-func (r *reconciler) reconcileGrafana(ctx context.Context) error {
+func (r *reconciler) reconcileGrafana(ctx context.Context) reconcileResult {
+	key := types.NamespacedName{
+		Name:      grafanaName,
+		Namespace: Namespace,
+	}
+
 	var grafana integreatlyv1alpha1.Grafana
-	key := types.NamespacedName{Name: grafanaName, Namespace: Namespace}
 	err := r.k8sClient.Get(ctx, key, &grafana)
+	if err != nil && !errors.IsNotFound(err) {
+		if err != nil {
+			r.logger.Info("Get failed" + string(errors.ReasonForError(err)))
+		}
 
-	if err == nil {
-		r.logger.Info("Updating Grafana instance")
-		grafana.Spec = newGrafana().Spec
-		return r.k8sClient.Update(ctx, &grafana)
+		if meta.IsNoMatchError(err) || errors.IsMethodNotSupported(err) {
+			r.logger.V(10).Info("Grafana CRD does not exist - NoMatchError")
+			return requeue(5*time.Second, nil)
+		}
+
+		return reconcileError(err)
 	}
 
+	// create
 	if errors.IsNotFound(err) {
 		r.logger.Info("Creating Grafana instance")
-		return r.k8sClient.Create(ctx, newGrafana())
+		err := r.k8sClient.Create(ctx, newGrafana())
+
+		// can fail because grafana operator hasn't created the CRD yet
+		if errors.IsNotFound(err) || errors.IsMethodNotSupported(err) || meta.IsNoMatchError(err) {
```

Comment:
I am still not too comfortable with this approach, I find it hard track what is going on because the logic of what the loop should do is scattered around. Why can't we make this check in the parent function and decide whether to requeue or not based on the error that is returned by the individual reconcilers (`reconcileGrafana`, `reconcileOperatorGroup`, etc)?

Reply:
I am not sure how it is going to change ... from the outer loop is is going to be like this 

```
1. reconcile
2. stop ? then return the result
3. else  next reconcile method
```
 
The inner loop / `reconcileFoobar` knows what errors can arise and how to handle then and if it needs to requeue or not - right?

Reply:
Alright, let's optimize this later if we have to. I have a feeling that we're missing something, but It doesn't have to be perfect from the get go.